### PR TITLE
Added support for Roidmi Eve

### PR DESCRIPTION
**Pull request by martin9000andersen**
I have added support for robot Roidmi Eve (Robot vacuum cleaner).

Many status values are implemented and many features are implemented e.g.:
    start-clean, stop, go-home, set_sound_volume, set_cleaning_mode, set_sweep_type, set_work_station_freq, start_dust, etc.

I do not know the syntax to activate cleaning of specific rooms, any input on how to make the function "start_room_sweep_unknown" work is highly appreciated.

### CODE REVIEW

File: miio/device.py
Comment (rytilahti) [inline miio/device.py:74]: Please rebase on top of the current master, #1083 should fix the device info problems.

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:3]:
```suggestion
```

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:165]:
```suggestion
class RoidmiVacuumStatus(DeviceStatus):
```
please extend this directly from the DeviceStatus for now. Also, please use the names and return types consistent to vacuumcontainers.py (e.g., `ConsumableStatus` for consumables), as we want to move slowly towards common interfaces among device types for interoperability.

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:174]: See my comment above, filter times should use timedeltas.

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:437]: This might be fixed by the recent #1075 - could you please confirm if that's the case?

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:359]: This doesn't currently seem to be queried at all. What does this indicate?

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:365]:
```suggestion
    def status(self) -> DeviceStatus:
```

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:120]: It is probably simpler to have this as a dictionary than as an enum (the value being a user-friendly message, see https://github.com/rytilahti/python-miio/blob/master/miio/vacuumcontainers.py#L16).

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:277]: Internal methods should be marked as such. Also, these could probably be moved to be in the scope of `forbid_mode`?

```suggestion
    def _parse_forbid_mode(self, val):
```

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:278]:
```suggestion
        def _secToClock(val):
```

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:289]: Instead of using json dumps, would it be possible to use the `DNDStatus` from `vacuumcontainers.py` for this? 

If not, please create a separate container class (derived from `DeviceStatus`).

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:310]:
```suggestion
```

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:318]:
```suggestion
```

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:524]:
```suggestion
    def set_dnd(self, start_hr: int, start_min: int, end_hr: int, end_min: int, active: bool = True)
```
This would make it similar to vacuum.py's `set_dnd` with the added `active` flag, for future proofing :-)

**Review by rytilahti — CHANGES_REQUESTED** [review]
Thanks for the PR! I did a first round of reviews and left some specific comments inline, but on generla level:
* Please add descriptive docstrings to all properties and methods belonging to the public API
* Remove commented out code (except for the variable mappings, in case someone wants to look into those later)
* Please add (at least) `is_on`, `is_paused`, `got_error` from https://github.com/rytilahti/python-miio/blob/master/miio/vacuumcontainers.py#L125 - the long term goal is to make all vacuum implementations to have a common interface, so this will be useful for that.

**Comment by martin9000andersen** [comment]
Thank you very much for your comments. I really like your long term goal of a common interface.
I will have a look at it in the coming days.

File: miio/device.py
Comment (martin9000andersen) [reply miio/device.py:74]: Yes, that commit fixed the problem

File: miio/roidmivacuum_miot.py
Comment (martin9000andersen) [reply miio/roidmivacuum_miot.py:437]: For Roidmi Eve there are two reasons for getting "checksum error": 1) is too many properties at a time (here I use max_properties), 2) Too many properties in total.
Commit #1075 did not affect No. 1, so I need to keep max_properties, but it did solved No. 2 (This means that I can uncomment all properties in MiotMapping). Thank you very much for telling me about this fix.

File: miio/roidmivacuum_miot.py
Comment (martin9000andersen) [reply miio/roidmivacuum_miot.py:359]: I am not sure. It is an Enum and can only have the value "Open". I will remove it from the public interface, and only leave an comment about this in MiotMapping.

**Comment by martin9000andersen** [comment]
I have updated the code to be more like vacuumcontainers.py (before it was closer to dreamevacuum_miot.py).

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:238]:
```suggestion
    def work_station_frequency(self) -> int:
```

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:329]:
```suggestion
    def lidar_enabled(self) -> bool:
```
Maybe?

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:596]:
```suggestion
    def set_timing(self, timing: str):
```
I think there is no need to describe this as unknown?

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:608]:
```suggestion
    def set_carpet_mode(self, auto_boost: bool):
```

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:681]: Could this be the name of the setting, e.g., "girl_en"?

File: miio/tests/test_roidmivacuum_miot.py
Comment (rytilahti) [inline miio/tests/test_roidmivacuum_miot.py:87]:
```suggestion
        assert status.charging_state == ChargingState(_INITIAL_STATE["charging_state"])
```
should be enough for all enum checks. If they are equal, their `repr()`s are also equal.

File: miio/tests/test_roidmivacuum_miot.py
Comment (rytilahti) [inline miio/tests/test_roidmivacuum_miot.py:76]:
```suggestion
```
This is unnecessary.

File: miio/tests/test_roidmivacuum_miot.py
Comment (rytilahti) [inline miio/tests/test_roidmivacuum_miot.py:122]:
```suggestion
    def test_consumable_status(self):
```
Avoid `CamelCasing`, it is also unnecessary to repeat "Roidmi" in tests.

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:687]: Please convert the actions to use `call_action(str)` and declare the actions in the mapping like: https://github.com/rytilahti/python-miio/pull/867/files#diff-2f2ea69d0a6bf409d623cea17e4744ce7b60739d4c29b79166a8c1979660bfb3R36-R42

The `call_action_by` is rather targeted for `miiocli` usage where there is no known symbol for it available.

**Review by rytilahti — COMMENTED** [review]
I added some more comments, I think it's good to go after this round but be aware that the API will change for the next major release to allow consolidating the common functionalities among all supported vacuums :-)

While doing the review I did some googling and started to wonder if this PR is for "Roidmi Eve Plus" or is there also a version without the "plus"? If that's the same one, this might be one option when it's time to upgrade away from my gen1 rockrobo.. What's your opinion about the device?

File: miio/roidmivacuum_miot.py
Comment (martin9000andersen) [reply miio/roidmivacuum_miot.py:329]: Text from the Mi App: "Priority use of radar - When turned on, the robot will use radar as the main detection sensor to help reduce collisions".
When this option is OFF, lidar is still used for mapping and navigation.
I do not think that "lidar_collision" or "lidar_enabler" are a great names. Do you have a better idea now that you know more about this option.

File: miio/roidmivacuum_miot.py
Comment (martin9000andersen) [reply miio/roidmivacuum_miot.py:681]: Yes, I think that "girl_en" can be part of the string. But I also think that it must be json format, and I do not know the json-key. (I have tried {"voice": "girl_en"} without lock)

**Comment by martin9000andersen** [comment]
> I added some more comments, I think it's good to go after this round but be aware that the API will change for the next major release to allow consolidating the common functionalities among all supported vacuums :-)

Thank you for the comments. I have added comments on a few points, the rest of the points look very clear and I hope to get time to solve them within a week.

**Comment by martin9000andersen** [comment]
> While doing the review I did some googling and started to wonder if this PR is for "Roidmi Eve Plus" or is there also a version without the "plus"? If that's the same one, this might be one option when it's time to upgrade away from my gen1 rockrobo.. What's your opinion about the device?

I have the "PLUS" model. The manual covers both models. The robot is the same, the "PLUS" is about the base. The "PLUS" base is with dust-collection. The manual shows a simple base for the non-PLUS model (but I have not seen the non-PLUS version for sale)
The miio protocol has a command to ask about "station-type", that is an int, for me this return 102, and I guess that the simple base will return a different number.

I am very satisfied with Roidmi Eve Plus. The largest draw back I see is missing Home assistant support, but with this pull request and the comming uniform vaccum interface, I think that will be solved.

File: miio/roidmivacuum_miot.py
Comment (martin9000andersen) [reply miio/roidmivacuum_miot.py]: This is a setting that you can change. I have changed the comment, and I have added a function to set this parameter

**Comment by martin9000andersen** [comment]
I have solved (most of) your comments, but also added a few extra functions.
Let me know if there are more changes you like me to make before merging.

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [reply miio/roidmivacuum_miot.py]: Okay, I'd suggest then using `led(self) -> bool` and `set_led(self, bool)` as in elsewhere (like in `airconditioner_miot.py` and `fan_miot.py` among others) to keep consistent naming!

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [reply miio/roidmivacuum_miot.py]: Okay, let's keep it this way, maybe someone figures it out at some point :-)

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [reply miio/roidmivacuum_miot.py]: I don't have a better recommendation for this at the moment, so let's keep it as it is.

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:289]:
```suggestion
    def is_water_box_attached(self) -> bool:
```
is used by rockrobo vacuums for exactly the same thing, right?

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:295]: Maybe?
```suggestion
    def dust_collection_frequency(self) -> int:
        """Frequency for emptying the dust bin.
        
            Example: 2 means base the dust bin is emptied every second cleaning.
        """
```

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:375]:
```suggestion
    def led(self) -> bool:
```
this will make it consistent with other integrations. It is enough to have the `station_` prefix below for the station led to differentiate it from this.

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:376]:
```suggestion
        """Return True if led/display on vaccum is on."""
```

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:415]: Is this `None` or `"None"` (as a string)? I'd suggest either removing this if it is not working as expected, as that would otherwise confuse anyone trying to use it (causing issue reports on github for a broken feature :-).

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:591]: Considering `roomstr` is not used at all, does something happen if this command gets executed?

If something happens, I'd move the FIXME to be `_LOGGER.warning()` to let users know about it (maybe someone finds out what the string should be) and otherwise remove the command decorator (so that it's not shown in the mioocli but left here just for a reference).

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:603]: Is this fixme still relevant considering the parameters are passed to the action?

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:627]:
```suggestion
    def set_led(self, on: bool):
```

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [inline miio/roidmivacuum_miot.py:658]: See my comment above about the getter for this.

File: miio/roidmivacuum_miot.py
Comment (martin9000andersen) [reply miio/roidmivacuum_miot.py:289]: On Roidme Eve the water-box is integrated in the dust-bin. The mop-module is approximately 1 cm high and can be attached below the combined dust+water-box.
Strictly speaking the mop is only a piece of fabric, and Roidmi Eve does not have a sensor to detect if that is pressent (and I guess that this is the same for Roborock).
If you would like a common name for Roborock and Roidme Eve, then we might be able to argue that the water-box on Roborock is a mop-module?
For now I think that best name for Roidme Eve it "is_mop_module_attached"?

File: miio/roidmivacuum_miot.py
Comment (martin9000andersen) [reply miio/roidmivacuum_miot.py:415]: I have now tried this again. The returned value is the same as if the feature is in not implemented. I will remove this function, and just leave a comment in the _MAPPING dict

File: miio/roidmivacuum_miot.py
Comment (martin9000andersen) [reply miio/roidmivacuum_miot.py:591]: The problem is that I have not figured out how to start an "action" with an argument.
I will comment out all the functions that are calling an action with an argument.

File: miio/roidmivacuum_miot.py
Comment (martin9000andersen) [reply miio/roidmivacuum_miot.py:603]: The problem is that I have not figured out how to start an "action" with an argument.
I will comment out all the functions that are calling an action with an argument.

File: miio/roidmivacuum_miot.py
Comment (rytilahti) [reply miio/roidmivacuum_miot.py:289]: Ah, so this is really about the mop, so I'd just use `is_mop_attached`.

**Review by rytilahti — APPROVED** [review]
All set, thanks @martin9000andersen!

**Comment by kali876** [comment]
I see the subject is a bit frozen :D 
I do not hide that it needs the functionality of vacuuming a specific zone.
If you know how to get mapping for zone vacuuming - I can help :)
I can follow all instructions as long as I get them :)
thx

**Pull request merged**
